### ISSUE TITLE
Dimension CreditCardNumber is not detected

### ISSUE THREAD

**Issue by jpslopes** [facebook/duckling#551]
Hello,

I tried to use all examples of credit card numbers from `corpus.hs` and none of them was detected as credit card number (it is detected as phone number when the credit card number has dashes, and it is detected as phone number and numeral/number when it has only digits).

So, I checked the dimensions and the `CreditCardNumber` is not registered neither in `commons.sh` nor in `EN.hs`. However, I noticed that the dimension `Numeral` is registered in both of them...

Thanks

Note: All tests were made using the Docker container and the locale `en_US`.

**Comment by chessai** [comment]
Looks like there are no rules for this in EN.

**Comment by chessai** [comment]
This shows that duckling is working for credit cards separated by dashes:

```
*Duckling.Debug> debug (makeLocale EN Nothing) "4111-1111-1111-1111" [Seal CreditCardNumber]
visa credit card number (4111-1111-1111-1111)
-- regex (4111-1111-1111-1111)
[Entity {dim = "credit-card-number", body = "4111-1111-1111-1111", value = RVal CreditCardNumber (CreditCardNumberValue {vNumber = "4111111111111111", vIssuer = Visa}), start = 0, end = 19, latent = False, enode = Node {nodeRange = Range 0 19, token = Token CreditCardNumber (CreditCardNumberData {number = "4111111111111111", issuer = Visa}), children = [Node {nodeRange = Range 0 19, token = Token RegexMatch (GroupMatch ["4111-1111-1111-1111"]), children = [], rule = Nothing}], rule = Just "visa credit card number"}}]
```

but not for those separated by spaces:

```
*Duckling.Debug> debug (makeLocale EN Nothing) "4111 1111 1111 1111" [Seal CreditCardNumber]
[]
```

**Comment by jpslopes** [comment]
Hello, I'm running a Docker container and sending POST HTTP requests to it. The request body is in `x-www-form-urlencoded` format. For example:
- locale: "en_US"
- text: "4111-1111-1111-1111"

And the response from the container is the following:
```json
[
    {
        "body": "4111-1111-1111-1111",
        "start": 0,
        "value": {
            "value": "4111111111111111",
            "type": "value"
        },
        "end": 19,
        "dim": "phone-number",
        "latent": false
    }
]
```

Another example (without dashes):
- locale: "en_US"
- text: "4111111111111111"

The response from Duckling:
```json
[
    {
        "body": "4111111111111111",
        "start": 0,
        "value": {
            "value": "4111111111111111",
            "type": "value"
        },
        "end": 16,
        "dim": "phone-number",
        "latent": false
    },
    {
        "body": "4111111111111111",
        "start": 0,
        "value": {
            "value": 4111111111111111,
            "type": "value"
        },
        "end": 16,
        "dim": "number",
        "latent": false
    }
]
```

I tried using all examples from corpus.sh and the output is the same for all of them.

**Comment by chessai** [comment]
> Hello, I'm running a Docker container and sending POST HTTP requests to it. The request body is in `x-www-form-urlencoded` format. For example:
> - locale: "en_US"
> - text: "4111-1111-1111-1111"
> 
> And the response from the container is the following:
> ```json
> [
>     {
>         "body": "4111-1111-1111-1111",
>         "start": 0,
>         "value": {
>             "value": "4111111111111111",
>             "type": "value"
>         },
>         "end": 19,
>         "dim": "phone-number",
>         "latent": false
>     }
> ]
> ```
> 
> Another example (without dashes):
> - locale: "en_US"
> - text: "4111111111111111"
> 
> The response from Duckling:
> ```json
> [
>     {
>         "body": "4111111111111111",
>         "start": 0,
>         "value": {
>             "value": "4111111111111111",
>             "type": "value"
>         },
>         "end": 16,
>         "dim": "phone-number",
>         "latent": false
>     },
>     {
>         "body": "4111111111111111",
>         "start": 0,
>         "value": {
>             "value": 4111111111111111,
>             "type": "value"
>         },
>         "end": 16,
>         "dim": "number",
>         "latent": false
>     }
> ]
> ```
> 
> I tried using all examples from `corpus.sh` and the output is the same for all of them.

Not at a computer right now, but have you tried passing `&dims="[\"credit-card-number\"]"` in addition to locale and text

**Comment by jpslopes** [comment]
OK, I did not know that it was possible to pass the dimensions in the request. I tried to pass the `CreditCardNumber` and it worked. So, the `CreditCardNumber` is not active by default?

**Comment by chessai** [comment]
> OK, I did not know that it was possible to pass the dimensions in the request. I tried to pass the `CreditCardNumber` and it worked. So, the `CreditCardNumber` is not active by default?

This is correct - it's not. This is more an issue of documentation. There should be clear documentation explaining this (would have cost you two minutes instead of a few days).

**Comment by jpslopes** [comment]
Is there some alternative to pass the dimensions in the request? For example, use an environment variable to configure Duckling when it starts executing to activate all dimensions. I'm asking this, because it looks to me that passing a list with all the dimensions in all requests can be heavy and have an impact on the performance.

**Comment by chessai** [comment]
> Is there some alternative to pass the dimensions in the request? For example, use an environment variable to configure Duckling when it starts executing to activate all dimensions. I'm asking this, because it looks to me that passing a list with all the dimensions in all requests can be heavy and have an impact on the performance.

No, there isn't currently. Keep in mind that the example application has no guarantee of "high" performance and is only intended to provide something that external users can use to test out duckling. So, I'm mostly against changing things too much for the sake of performance. If you are considering using duckling in some performance-intensive context, I would suggest rolling something on your own.

That being said, for the example application, the change to accept dimensions through an env var is so simple to implement/understand/maintain that I'm not worried about the overhead. If you are interested in doing this, I would accept a PR.  In case you are unfamiliar with Haskell, you would need to start with looking at [lookupEnv](https://hackage.haskell.org/package/base-4.14.0.0/docs/System-Environment.html#v:lookupEnv).

**Comment by chessai** [comment]
Actually, I think that all should just be enabled by default. And passing &dims will just specify if you want a subset.

**Comment by chessai** [comment]
closed by 17f11135f2ccbf4cb89186dd5416a237d6d6a928

now example application should do every available dimension by default. the new behavior is documented

**Comment by jpslopes** [comment]
OK, thank you for the update.

> Looks like there are no rules for this in EN.
> 
> EDIT: CreditCardNumber is in the common rules, so should be present regardless of locale.
> 
> How are you invoking `duckling`?

I did not notice this edit. So, I looked up in `Dimensions/Common.hs` and noticed that `CreditCardNumber` is not present there. I tried adding it to the list of common dimensions and it worked like a charm.
```haskell
allDimensions :: [Seal Dimension]
allDimensions =
  [ Seal AmountOfMoney
  , Seal CreditCardNumber
  , Seal Email
  , Seal Numeral
  , Seal PhoneNumber
  , Seal Url
  ]
```
When sending a request with the following input (without passing the parameter `dims`):
- text: "4111-1111-1111-1111"
- locale: ""en_US

The response was the following:
```json
[
    {
        "body": "4111-1111-1111-1111",
        "start": 0,
        "value": {
            "value": "4111111111111111",
            "issuer": "visa"
        },
        "end": 19,
        "dim": "credit-card-number",
        "latent": false
    },
    {
        "body": "4111-1111-1111-1111",
        "start": 0,
        "value": {
            "value": "4111111111111111",
            "type": "value"
        },
        "end": 19,
        "dim": "phone-number",
        "latent": false
    }
]
```
Note: This test was performed using the Duckling's code updated on Nov 12, 2020.
